neqo-http3/src/connection_client.rs

```diff
@@ -2535,4 +2360,569 @@ mod tests {
+                    }
+                }
+                ConnectionEvent::SendStreamWritable { stream_id } => {
+                    assert!(matches!(stream_id, 2 | 6 | 10));
+                }
+                ConnectionEvent::StateChange(_) => (),
+                _ => panic!("unexpected event"),
+            }
+        }
+        assert!(control_stream);
+        assert!(qpack_encoder_stream);
+        assert!(qpack_decoder_stream);
+        assert_eq!(request, expect_request);
+    }
+
+    fn exchange_token(client: &mut Http3Client, server: &mut Connection) -> Vec<u8> {
+        server.send_ticket(now(), &[]).expect("can send ticket");
+        let out = server.process_output(now());
+        assert!(out.as_dgram_ref().is_some());
+        client.process_input(out.dgram().unwrap(), now());
+        assert_eq!(client.state(), Http3State::Connected);
+        client.resumption_token().expect("should have token")
+    }
+
+    fn start_with_0rtt() -> (Http3Client, TestServer) {
+        let (mut client, mut server) = connect();
+        let token = exchange_token(&mut client, &mut server.conn);
+
+        let mut client = default_http3_client();
+
+        let server = make_default_server();
+
+        assert_eq!(client.state(), Http3State::Initializing);
+        client
+            .set_resumption_token(now(), &token)
+            .expect("Set resumption token.");
+
+        assert_eq!(client.state(), Http3State::ZeroRtt);
+
+        (client, server)
+    }
+
+    #[test]
+    fn zero_rtt_negotiated() {
+        let (mut client, mut server) = start_with_0rtt();
+
+        let out = client.process(None, now());
+
+        assert_eq!(client.state(), Http3State::ZeroRtt);
+        assert_eq!(*server.conn.state(), State::WaitInitial);
+        let out = server.conn.process(out.dgram(), now());
+
+        // Check that control and qpack streams are received and a
+        // SETTINGS frame has been received.
+        // Also qpack encoder stream will send "change capacity" instruction because it has
+        // the peer settings already.
+        check_control_qpack_request_streams_resumption(
+            &mut server.conn,
+            ENCODER_STREAM_DATA_WITH_CAP_INSTRUCTION,
+            false,
+        );
+
+        assert_eq!(*server.conn.state(), State::Handshaking);
+        let out = client.process(out.dgram(), now());
+        assert_eq!(client.state(), Http3State::Connected);
+
+        let _out = server.conn.process(out.dgram(), now());
+        assert_eq!(*server.conn.state(), State::Connected);
+
+        assert!(client.tls_info().unwrap().resumed());
+        assert!(server.conn.tls_info().unwrap().resumed());
+    }
+
+    #[test]
+    fn zero_rtt_send_request() {
+        let (mut client, mut server) = start_with_0rtt();
+
+        let request_stream_id = client
+            .fetch("GET", "https", "something.com", "/", &[])
+            .unwrap();
+        assert_eq!(request_stream_id, 0);
```

Comment:
It might pay to factor this into a function as well.  It's the standard request you make, you have its encoding up above, but you rely on copy-paste to get it right.